### PR TITLE
interactive: refactor to one use of get_all_passes/get_all_dialects

### PR DESCRIPTION
**Pull request by alexarice**
Changes the interactive app object to store a pass list and context, which are set for now in the initialization function. This will allow further refactors which allow these to be overridden more easily by downstream projects.

### CODE REVIEW

File: xdsl/interactive/get_all_available_passes.py
Comment (superlopuh) [inline xdsl/interactive/get_all_available_passes.py:18]: why pass the ctx here?

Comment (alexarice) [reply xdsl/interactive/get_all_available_passes.py:18]: It's needed for the parser

File: xdsl/interactive/app.py
Comment (superlopuh) [inline xdsl/interactive/app.py:94]: I think it's quite important to not add mutable state to the app, could you please refactor this to avoid passing the context? Probably something like this:
```suggestion
    all_dialects: tuple[tuple[str, Callable[[], Dialect]]]
```
And just pass that in the init. Similarly, `all_passes` below should be passed in the init. This will also massively improve our tests, as we can initialize the test app with a smaller set of dialects and passes for it to condense.

Comment (alexarice) [reply xdsl/interactive/app.py:94]: This seems like a good idea. I'm leaning towards making both "all_passes" and "all_dialects" into dictionaries. I don't really see why they should be tuples of tuples?

File: tests/interactive/test_app.py
Comment (alexarice) [inline tests/interactive/test_app.py:45]: Just occurred to me this might have been a joke that I wasn't funny enough to understand

**Comment by alexarice** [comment]
@superlopuh, I've had a second pass at this given the feedback, let me know what you think.

File: xdsl/interactive/app.py
Comment (superlopuh) [reply xdsl/interactive/app.py]: because tuples are frozen, mostly. a frozendict would also work, but requires an extra dependency

File: xdsl/interactive/passes.py
Comment (superlopuh) [inline xdsl/interactive/passes.py:10]: the immutability is why this was a tuple also

Comment (alexarice) [reply xdsl/interactive/passes.py:10]: Ah, I assumed it was for the ordered Ness (python dicts are now ordered). Happy to make both tuples or make them private or both

**Comment by alexarice** [comment]
Reverted to using tuples for both dialects and passes

File: xdsl/interactive/app.py
Comment (superlopuh) [inline xdsl/interactive/app.py:159]: This is code smell to me, if someone already has the dialects and passes as a tuple, they would need to first build a dict to pass to the init, it's a shame to make it impossible to do the dumb/fast thing.

Comment (alexarice) [reply xdsl/interactive/app.py:159]: I kind of get this, though to me the dictionary is the "correct" interface, in that it stops you assigning the same name to two different dialects/passes, whereas storing it as a tuple is really just an implementation detail.

File: xdsl/interactive/app.py
Comment (alexarice) [reply xdsl/interactive/app.py:141]: I have no idea how this happened

File: xdsl/interactive/app.py
Comment (alexarice) [reply xdsl/interactive/app.py:160]: Note that `all_passes` has to have every pass actually called at some point as well. Unsure what the nicest interface is, realistically this is called once per project

Comment (superlopuh) [reply xdsl/interactive/app.py:160]: I agree in general, and this is where frozendict is nice, as it guarantees both unicity and frozenness, but in this case I think we'll be ok with the potential for errors in this way. In either case, this is a separate discussion from whether the init allows for passing the object to be stored directly or not

Comment (superlopuh) [reply xdsl/interactive/app.py:160]: Not dictionary, since it's mutable, so tuple for now, please. In theory, we could add a dependency to frozendict, but I'm not sure it's worth it for this PR.

Comment (alexarice) [reply xdsl/interactive/app.py:160]: I meant just for the init function

Comment (superlopuh) [reply xdsl/interactive/app.py:160]: Ah, I don't really see the point, if it's just `tuple(all_dialects.items())` on the client side, and avoids `isinstance` checks in the init.

Comment (superlopuh) [reply xdsl/interactive/app.py:160]: I had a secret plan of making that change after you merge your PR :)

Comment (superlopuh) [reply xdsl/interactive/app.py:160]: Because it seems like too big a change to do both at the same time, but there's not a particularly good reason to make it eager at the app level.

Comment (alexarice) [reply xdsl/interactive/app.py:160]: What's the solution then? 
- dictionaries with default and convert in init
- choice of dictionary/tuple and need isinstance checks
- only tuple but don't give defaults

Comment (superlopuh) [reply xdsl/interactive/app.py:160]: only tuple, don't give defaults

Comment (alexarice) [reply xdsl/interactive/app.py]: Done, tests aren't pretty atm but I'm guessing you're hoping fine tune the passes they need in a future pr

Comment (alexarice) [reply xdsl/interactive/app.py]: Realised I could have put the default for dialects, as this shouldn't be too expensive, let me know if you want it in

**Review by superlopuh — APPROVED** [review]
Thank you!

File: xdsl/interactive/app.py
Comment (superlopuh) [reply xdsl/interactive/app.py]: I think the user needs to know they're doing something potentially very expensive, and to expose how to customize it for downstream users such as yourself

**Comment by alexarice** [comment]
Rebased past conflict, will merge if ci is green

**Pull request merged**
